server/api/messaging.py

```diff
@@ -0,0 +1,207 @@
+    # Get the conversation between the two users.
+    conversation = Conversation.get_conversation(user, other_user)
+    # Messages in the conversation are already ordered by descending created_at (date/time)
+    for message in conversation.messages.all():
+        if message.sender != user:
+            profile_image = message.sender.profile.profile_image
+        else:
+            profile_image = message.receiver.profile.profile_image
+        message_dict = {
+            'sender': message.sender.profile.name,
+            'image_url': profile_image,
+            'content': message.content
+        }
+        for field, value in message.get_formatted_info:
+            message_dict[field] = value
+        response_dict['messages'].append(message_dict)
+    response_dict['status'] = 200
+    return jsonify(response_dict), 200
+
+@messaging_views.route("/get_conversation_previews", methods=["GET"])
+@authenticate
+def get_convo_previews():
+    """
+    Get all conversations this user has, then build a
+    dictionary where the outer most key is 'messages'
+    and its value is a list of dictionaries. The fields
+    for each message in the dictionaries are...
+    1. The name of the other user (at the key of 'name')
+    2. The other users profile image (at the key of 'image_url')
+    3. The last message sent in the conversation (at the key of 'content')
+    """
+    response_dict = {'messages': []}
+    # If we're getting the previews for the ConversationView
+    # there will be an 'exclude' key with the value of a
+    # user id of the other user in the current conversation
+    request_dict = request.get_json()
+    excluded_user = None
+    if request_dict.get('user_id', None) != None:
+        excluded_user = User.get_instance(**{'id': request_dict.get('user_id')})
+    base_query = session.query(Conversation).filter(or_(
+            user_one=user,
+            user_two=user
+        ))
+    if excluded_user != None:
+        conversations = base_query.filter(
+            and_(
+                user_one!=excluded_user,
+                user_two!=excluded_user
+            )
+        ).all()
+    else:
+        conversations = base_query.all()
+    if len(conversations) == 0:
+        response_dict['status'] = 204
+        response_dict['message'] = "No conversations yet"
+        return jsonify(response_dict), 204
+    for conversation in conversations:
+        newest_message = conversation.messages.first()
+        sender_profile = newest_message.sender.profile
+        if sender_profile.name != user.profile.name:
+            other_user_profile = sender_profile.name
+        else:
+            other_user = newest_message.receiver.profile.name
+        message_info = {
+            'name': other_user_profile.name,
+            'image_url': other_user_profile.profile_image,
+            'content': newest_message.content,
+            'created_at': newest_message.get_formatted_time
+        }
```

Comment:
This is good!
One note would be that it's best not to use `name` to check which profile image to use. In production systems you'll quickly find that anything other than object Ids you can't rely on being unique lol.

Reply:
LOL I can't believe I didn't think about that. Thank you, I'll change it to check the ids instead lol